test/unittests/CandidateTest.php

```diff
@@ -658,6 +658,46 @@ public function testGetNextVisitNoReturnsOneWhenNextVisitDoesNotExist()
         $this->_candidate->select($this->_candidateInfo['CandID']);
         $this->assertEquals(1, $this->_candidate->getNextVisitNo());
     }
+    /**
+     * Test getAgeYears returns age as int years
+     *
+     * @covers Candidate::getAgeYears()
+     * @return void
+     */
+    public function testGetAgeYearsReturnsIntYears()
```

Comment:
@johnsaigle would you mind looking at these test functions? I'm not quite sure what I'm doing here. do you also have any insights as to how to write the test function for getAge()?

Reply:
The current unit tests look good!

I did a bit research on unit testing functions that refer to the current date "now" which is a little tricky because unit tests rely on a static value. For example we would want to write a test that, given a candidate born on e.g. January 1st, 2019, the `getAge()*` functions would return 1 year, 13 months, or 365 + 24 days depending on what date interval you care about. (I'm writing this on Feb. 25th 2020). But if we hardcode that then we can't test for 365 + 24 days because that will change each day...

The way to approach this is by changing the function getAge() to this:

```php
    public function getAge($referenceDate = new DateTime()): DateInterval
    {
        $dob = new DateTime($this->candidateInfo["DoB"]);

        return $dob->diff($referenceDate);
    }
```

This allows us to choose a moment from which to measure the candidate's age. It means adding an optional parameter to getAge() but we don't have to use it in the actual codebase.

Doing this means that when the unit test is run we can calculate it based on a fixed date and ensure that the calculation is correct. 

The unit test would look something like this:

```php
public function testGetAgeReturnsCorrectValues() {
    $referenceDate = new DateTime('2020-02-25');
    // For a candidate with birthday == '2019-01-01'
    assertsEquals($this->_candidate->getAgeInYears(), 1);
    assertsEquals($this->_candidate->getAgeInMonths(), 13);
    assertsEquals($this->_candidate->getAgeInDays(), 365 + 24);
}
```

Reply:
The current unit tests look good!

I did a bit research on unit testing functions that refer to the current date "now" which is a little tricky because unit tests rely on a static value. For example we would want to write a test that, given a candidate born on e.g. January 1st, 2019, the `getAge()*` functions would return 1 year, 13 months, or 365 + 24 days depending on what date interval you care about. (I'm writing this on Feb. 25th 2020). But if we hardcode that then we can't test for 365 + 24 days because that will change each day...

The way to approach this is by changing the function getAge() to this:

```php
    public function getAge($referenceDate = new DateTime()): DateInterval
    {
        $dob = new DateTime($this->candidateInfo["DoB"]);

        return $dob->diff($referenceDate);
    }
```

This allows us to choose a moment from which to measure the candidate's age. It means adding an optional parameter to getAge() but we don't have to use it in the actual codebase.

Doing this means that when the unit test is run we can calculate it based on a fixed date and ensure that the calculation is correct. 

The unit test would look something like this:

```php
public function testGetAgeReturnsCorrectValues() {
    $referenceDate = new DateTime('2020-02-25');
    // For a candidate with birthday == '2019-01-01'
    assertsEquals($this->_candidate->getAgeInYears(), 1);
    assertsEquals($this->_candidate->getAgeInMonths(), 13);
    assertsEquals($this->_candidate->getAgeInDays(), 365 + 24);
}
```